ENH: Read arbitrary bit depth (including 24-bit) WAVs

endolith opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Fixes https://github.com/scipy/scipy/issues/1930

Adapted from part of https://github.com/scipy/scipy/pull/6852 and https://github.com/scipy/scipy/issues/6849

#### What does this implement/fix?
This adds support for reading arbitrary bit depth integer PCM WAV files, from 1- to 32-bit (including the commonly-requested 24-bit).  

I moved the bit depth check inside the read data function, since it's different for float vs integer formats.

As far as I can tell, the only 64-bit WAV files in existence are 64-bit float, not 64-bit integer PCM, so integer format stops at 32-bit.  The WAV fmt chunk could theoretically allow for 65,536-bit files, but I think 32 is sufficient. :)

I also doubt that float16 WAV files exist, but it's currently there.  

As far as I know, memmap can't handle anything that isn't already in a numpy native data type, so that raises exceptions for bit depths  (such as 24-bit)  that would require a data type other than {uint8, int16, int32}:

| Bit depth | Container bytes | mmap=False | mmap=True |
| --------- | --------------- | ---------- | --------- |
| 1 to 8    | 1               | uint8      | uint8     |
| 9 to 16   | 2               | int16      | int16     |
| 17 to 24  | 3               | int32      | No        |
| 25 to 32  | 4               | int32      | int32     |

This outputs data in the same left-justified format used in the WAV file itself, which is backwards compatible with the existing behavior and compatible with memmap.  So a 6-bit file will have a max (unsigned) value of 0b1111_1100 = 252 for instance, and 24-bit would have a max (signed) value of 0b0111_1111_1111_1111_1111_1111_0000_0000 = +2,147,483,392

Examples of unusual bit depths:

* [Adobe Audition's 20-bit format](https://flirtfm.ie/volunteers/downloads/audition_3_help.pdf#page=256)
* [M1F1-int12-AFsp.wav](http://www-mmsp.ece.mcgill.ca/Documents/AudioFormats/WAVE/Samples/AFsp/M1F1-int12-AFsp.wav) 12-bit example
* [LTspice can generate WAVs](http://ltwiki.org/LTspiceHelp/LTspiceHelp/_WAVE_Write_selected_nodes_to_a_wav_file_.htm) of simulated electrical signals with 1- to 32-bit depth ([example generator](https://gist.github.com/endolith/b3d4cc7a7d8d7f400cc9b542f5727691))

I wrote "However, if a file actually contains data past its specified bit depth, those bits will be read and preserved.)" because Audition documentation mentions this happening:

> The extra 4 bits are actually the remaining valid bits when saving, and they are used when reading (thus still giving 24-bit accuracy if those bits were actually present when writing). Applications either fill those last 4 bits with zeros or with actual data

I would also like to add other output types for integer PCM, but that can be another PR:

* Right-justified format: Signed int for all bit depths (even <=8), also in the smallest numpy type that will fit, right-justified, so that 6-bit max is 0b00111111 = 63, for instance.
* Float "full-scale" normalization: Convert the data to float by dividing by 2**(N-1)-1 (https://github.com/scipy/scipy/issues/12059)
* Float "fixed point" normalization: Convert the data to float by dividing by 2**(N-1)

None of these would be compatible with mmap because they need to be copies (though float WAV could still be mapped to float directly)

(Also to return a metadata object that includes bit depth, or an info() function that returns it)




#### Additional information
3-byte code is based on https://github.com/perimosocordiae/scipy/commit/9dba3f35258273fd81bf9bf7ea030c6e14780389
which is based on https://gist.github.com/josephernest/3f22c5ed5dabf1815f16efa8fa53d476
which is based on https://github.com/WarrenWeckesser/wavio/blob/master/wavio.py#L60

perimosocordiae commented on scipy/io/wavfile.py:
```diff
@@ -419,21 +419,42 @@ def _read_data_chunk(fid, format_tag, channels, bit_depth, is_big_endian,
+                             f"has {bit_depth}-bit data.")
+    else:
+        # This should have been caught by _read_fmt_chunk already
+        raise ValueError(f"Unknown wave file format")
```
Missing a {format_tag} here.

endolith replied:
Yeah, that checking is already covered inside the read fmt chunk function, so this should never be reached, so I wasn't sure what to do here.  Duplicate the same code from the read fmt function?

```
        try:
            format_name = WAVE_FORMAT(format_tag).name
        except ValueError:
            format_name = f'{format_tag:#06x}'
        raise ValueError(f"Unknown wave file format: {format_name}. Supported "
                         "formats: " +
                         ', '.join(x.name for x in KNOWN_WAVE_FORMATS))
```

larsoner replied:
maybe a little helper `_bad_format(format_name)` so that there is less code dup?

endolith commented:
One test failure is from testing 64-bit int format, which I don't think is actually used, but theoretically could be.  Should we support that (as well as the 5, 6, 7-byte formats in-between, that are even less likely to be used) or remove it from the tests?

Also `test_write_roundtrip` currently generates 20 combinations only to reject 12 of them.  It generates {i1, i2, i4, i8, u1, u2, u4, u8, f1, f2, f4, f8, g1, g2, g4, g8, q1, q2, q4, q8} but only keeps {u1, i2, i4, i8, f4, f8, g, q}.  

I'm not sure what the interface for the `write` function should be, but if it's going to support arbitrary bit depth, too, then we need to test a few different bit depths (but not all of them), and only {32, 64} float bit depths.  So it should probably be an explicit list instead of rejecting many things.

larsoner commented:
> Should we support that (as well as the 5, 6, 7-byte formats in-between, that are even less likely to be used) or remove it from the tests?

I would proceed incrementally rather than try to be as complete as possible in one PR. If 24-bit is commonly requested, I'd rather get support for it soon and add support for other way less common (or even nonexistent) formats later.

> Also test_write_roundtrip currently generates 20 combinations only to reject 12 of them. It generates {i1, i2, i4, i8, u1, u2, u4, u8, f1, f2, f4, f8, g1, g2, g4, g8, q1, q2, q4, q8} but only keeps {u1, i2, i4, i8, f4, f8, g, q}.

Probably worth just enumerating the used ones directly in this case

endolith commented:
> I'd rather get support for it soon and add support for other way less common (or even nonexistent) formats later.

Well it would actually be _removing_ support for 64-bit int, since that's tested for.   The tests actually exclude float16 from testing, though, so I will remove that from supported formats.

larsoner commented:
> Well it would actually be removing support for 64-bit int, since that's tested for.

Can we keep the 64-bit int in that case?

endolith commented:
> Can we keep the 64-bit int in that case?

Yes, but then do we also need to write rearranging code to handle everything in between?  33-63 bit are 5, 6, or 7 byte and need to be stored in int64, too.

larsoner commented:
I think we have to stay pragmatic, how about "only support what maps directly onto a NumPy dtype, or is used widely enough to justify the additional maintenance burden"?  This means in64 stays because it's trivial, int24 stays and we put in extra effort for because it's widely used, but we don't bother with int33 because nobody in practice really uses it?

endolith commented:
Ok I modified it to (hopefully) support anything from 1-64 bit

endolith commented:
Also the documentation saying "32-bit floating-point" and "32-bit PCM" might be slightly wrong, since both are actually considered PCM?

endolith commented:
So the float128 errors are from the `g` dtype, and presumably also where the https://github.com/scipy/scipy/pull/6184#issuecomment-636357138 errors came from. https://numpy.org/devdocs/reference/arrays.scalars.html says this is longfloat, which is of platform-dependent length (float64 for me).

Should we really be testing/supporting platform-dependent types for output to WAV files?  I don't think anyone uses anything other than 32-bit and 64-bit float, or that reading them is supported by most software.

larsoner commented:
Okay with me to get rid of 128 and 96 float, and to update the docstring regarding PCM int/float

X-Raym commented:
Very nice, look for this to be implemented !

endolith commented:
It's failing because of this line 

    grep -rlIP '[^\x00-\x7F]' scipy | grep '\.pyx\?' | sort > unicode.out; grep -rlI '# -\*- coding: \(utf-8\|latin-1\) -\*-' scipy | grep '\.pyx\?' | sort > coding.out; comm -23 unicode.out coding.out > test_code.out; cat test_code.out;  test \! -s test_code.out

in https://github.com/scipy/scipy/blob/master/.travis.yml

which is some kind of unicode filter?  but it's not clear what it's doing or what's allowed.  Can this test be documented in some way?

WarrenWeckesser commented:
The quick fix is to change "µ-law" to "mu-law".

WarrenWeckesser commented:
That check for characters outside the range 0-127 with no encoding specified was added in https://github.com/scipy/scipy/pull/9615, in response to https://github.com/scipy/scipy/issues/9606.  We can probably remove the check; I'll open a new issue about it.

If you want to leave "µ-law" in the file, then until the check is removed, it looks like a work-around is to add an explicit encoding comment at the beginning of the file, using the format expected by the regex in the check:

    # -*- coding: utf-8 -*-

endolith commented:
So I updated the documentation, made the `_raise_bad_format`, added a bunch of test files (<1 kB total), and all tests pass except `test_arpack.py:test_symmetric_modes` which I assume is unrelated.  It should now read any integer WAV file from 1-64 bits, and float WAV files of 32 or 64 bits.  Any other changes I would leave for other PRs.

I'm going to add some more comments, but otherwise I think I'm done with this PR.

endolith commented:
> If you want to leave "µ-law" in the file, then until the check is removed, it looks like a work-around is to add an explicit encoding comment at the beginning of the file, using the format expected by the regex in the check:
>     # -*- coding: utf-8 -*-


Isn't that the default encoding in Python 3 though?

endolith commented on scipy/io/wavfile.py:
```diff
@@ -419,21 +423,42 @@ def _read_data_chunk(fid, format_tag, channels, bit_depth, is_big_endian,
+
+    if format_tag == WAVE_FORMAT.PCM:
+        if 1 <= bit_depth <= 8 or bytes_per_sample in {3, 5, 6, 7}:
+            dtype = 'u1'
```
Actually, instead of adding comments here explaining that `u1` represents raw bytes, it should just use `V1` type.

WarrenWeckesser commented:
@endolith, yes, the default encoding for Python 3 source files is UTF-8.  The history of the existing check for non-ASCII characters can be found in the github issue and PR that I linked to earlier.  The check is a holdover from supporting Python 2.  Now that we support only Python 3, the check is not needed--at least as far as it concerns ensuring that Python works.  Why we might still keep the check in one form or another (and whether or not the check should be overridden like it is now if an explicit encoding is given in a comment in the Python file) should be discussed in https://github.com/scipy/scipy/issues/11925.

endolith commented on scipy/io/tests/test_wavfile.py:
```diff
@@ -85,6 +85,221 @@ def test_read_5():
+        assert_equal(data[0, 0], 128)  # Midpoint is 128 for <= 8-bit
+        assert_equal(data.min(), 0)  # Lowest possible
+
+    del data
```
Yeah I just copied them from existing tests.  They were added along with mmap support: https://github.com/scipy/scipy/commit/975cb7f55c

But garbage collection should handle this automatically even for mmap right?

endolith replied:
@pv Can you comment on why these were added?

pv replied:
You should generally drop references to buffers to mmapped files, before closing them (or deleting, on Windows). But it seems many of those were added also for non-mmap files.

endolith replied:
Ok but it's not necessary for `test_read_incomplete_chunk` or `test_read_early_eof`?

endolith replied:
Oh because it's the data array that needs to be deleted, not the file object.

endolith commented on scipy/io/wavfile.py:
```diff
@@ -419,21 +423,47 @@ def _read_data_chunk(fid, format_tag, channels, bit_depth, is_big_endian,
             dtype = f'{fmt}i{bytes_per_sample}'
         else:
+            raise ValueError("Unsupported bit depth: the wav file "
+                             f"has {bit_depth}-bit data.")
```
Ok, fixed all those and everything passes now

sethtroisi commented on scipy/io/wavfile.py:
```diff
@@ -421,25 +425,53 @@ def _read_data_chunk(fid, format_tag, channels, bit_depth, is_big_endian,
+        elif bytes_per_sample in {3, 5, 6, 7}:
+            # No compatible dtype.  Load as raw bytes for reshaping later.
+            dtype = 'V1'
+        elif bit_depth <= 64:
```
Would this be more clear as `elif bytes_per_sample in (2,4,8):`?

endolith replied:
Hmmm

- Whether the data is signed vs unsigned is determined by the bit depth being 8 or less bits. 
- Whether the data corresponds to a native dtype vs needs reshaping is determined by the container byte size.

These aren't necessarily related, either.  A WAV file could store unsigned 5-bit data in 4-byte containers, for example, which ...

Wait, this doesn't work, does it?  If a WAV stores 8-bit data in a 3-byte container, then we need to interpret it as unsigned, because it's <= 8-bit, but reshape it to fit in 4-byte container, since there's no uint24, so it should be reshaped to dtype=uint32?

http://www-mmsp.ece.mcgill.ca/Documents/AudioFormats/WAVE/Docs/riffmci.pdf does say "The size of i is the smallest number of bytes required to contain the specified sample size." so I guess that's not legal (and probably doesn't exist anyway).

and http://www-mmsp.ece.mcgill.ca/Documents/AudioFormats/WAVE/WAVE.html describes a real-world format with 24 bit depth but 4 byte container size which is actually float stored as int.  So we wouldn't want to open that as int.  

So it should only accept files that have a container size that doesn't exceed the bit depth?  And if someone wants to store 24-bit data and read it back mmapped, they can store it as right-shifted int32?  (Again, most of these corner cases will never be seen by anyone.)

sethtroisi commented on scipy/io/wavfile.py:
```diff
@@ -499,52 +531,70 @@ def read(filename, mmap=False):
-    32-bit PCM             -2147483648  +2147483647  int32
-    16-bit PCM             -32768       +32767       int16
-    8-bit PCM              0            255          uint8
+    32-bit integer PCM     -2147483648  +2147483647  int32
```
Is 64-bit floating-point supported? L442 makes me suspect yes

endolith replied:
Yes but I don't think any such files exist in the real world.  They certainly aren't common

sethtroisi replied:
ahh, I read that comment in another place but didn't put 2 and 2 together.

endolith commented:
The test failure seems unrelated: test_symmetric_modes in arpack

endolith commented:
Anything else I need to do on this?  Any other opinions on https://github.com/scipy/scipy/pull/12287#discussion_r459191917 ?

sethtroisi commented:
@larsoner  +1 from me. I reviewed with 70% rigor a while ago.

larsoner commented:
Thanks @endolith !

X-Raym commented:
This will helps a lot of people, thanks ! @endolith
